foundationdb/src/tuple/subspace.rs

```diff
@@ -25,24 +25,37 @@ use std::hash::Hash;
 #[derive(Debug, Clone, Eq, PartialEq, Hash)]
 pub struct Subspace {
     prefix: Vec<u8>,
+    versionstamp_offset: VersionstampOffset,
 }
 
 impl<E: TuplePack> From<E> for Subspace {
     fn from(e: E) -> Self {
-        Self { prefix: pack(&e) }
+        let mut prefix = vec![];
+        let versionstamp_offset = pack_into(&e, &mut prefix);
+        Self {
+            prefix,
+            versionstamp_offset,
+        }
     }
 }
 
 impl Subspace {
     /// `all` returns the Subspace corresponding to all keys in a FoundationDB database.
     pub fn all() -> Self {
-        Self { prefix: Vec::new() }
+        Self {
+            prefix: Vec::new(),
+            versionstamp_offset: VersionstampOffset::None { size: 0 },
+        }
     }
 
     /// Returns a new Subspace from the provided bytes.
     pub fn from_bytes(prefix: impl Into<Vec<u8>>) -> Self {
+        let prefix = prefix.into();
         Self {
-            prefix: prefix.into(),
+            versionstamp_offset: VersionstampOffset::None {
```

Comment:
Here we assume, `prefix` doesn't contains any versionstamp.

There is two option:
 - we assume `prefix` in from_bytes won't and will not contains any versionstamp element (ie. raw), and continue with this pr logic (no breaking change)
 - we assume a Subspace is always encoded (ie. unpack is always possible), then we don't need to add a `versionstamp_offset` member, and can compute the offset when needed by unpacking the prefix (breaking change) (I did use a custom encoding in one of my project, but at that point I was also not relying on Subspace nor the tuple module).

Reply:
There's a third option which actually would make things harder to get wrong too.

It's quite possible to have create a tuple with multiple versionstamps now which will be a runtime error if we try pack it.

It's also possible to call the wrong pack command on an incomplete versionstamp and get an invalid entry in foundationdb silently.

An even more breaking change could be to introduce a versionstamp-aware encoding to make this more strict and not even allow misusing versionstamps.

something like

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0e60959fe95209bb4c5be18afc978855

I'd be happy to introduce something like this as a `tuple_v2` module.

Reply:
I quite like this idea.

Reply:
I experimented with this a little bit to make a generic packing implementation but it falls short when needing to implement it for the tuple types, where the transitive property doesn't work here:

https://github.com/DataDog/foundationdb-rs/blob/4637bd8bfd12111720532bc13868518c9c36f497/foundationdb/src/tuple/pack_v2.rs#L333-L357

I think it might be too complicated to do to match the existing implementation, but have a look if you like.